app/components/user/TwoFactor/RecoveryCodes/RecoveryCodeDialog.js

```diff
@@ -0,0 +1,206 @@
+// @flow
+
+import * as React from 'react';
+import { createFragmentContainer, graphql, commitMutation } from 'react-relay/compat';
+import { CopyToClipboard } from 'react-copy-to-clipboard';
+import GraphQLErrors from 'app/constants/GraphQLErrors';
+import Button from 'app/components/shared/Button';
+import Dialog from 'app/components/shared/Dialog';
+import RecoveryCodeList from 'app/components/RecoveryCodeList';
+import type { RecoveryCodeDialog_totp } from './__generated__/RecoveryCodeDialog_totp.graphql';
+import type { _RecoveryCodeDialogCreateMutationVariables, RecoveryCodeDialogCreateMutationResponse } from './__generated__/RecoveryCodeDialogCreateMutation.graphql';
+
+type RecoveryCodes = $PropertyType<RecoveryCodeDialog_totp, 'recoveryCodes'>;
+
+type Props = {
+  onRequestClose: Function,
+  totp: RecoveryCodeDialog_totp,
+  relay: Object
+};
+
+type State = {
+  generatingRecoveryCodes: boolean,
+  generatedRecoveryCodes: boolean,
+  recoveryCodes: RecoveryCodes
+};
+
+function recoveryCodeText(recoveryCodes: RecoveryCodes): ?string {
+  if (recoveryCodes && recoveryCodes.codes) {
+    return recoveryCodes.codes.reduce((memo, { code }) => memo.concat(code), []).join('\n');
+  }
+  return '';
+}
+
+class RecoveryCodesDialog extends React.PureComponent<Props, State> {
+  state = {
+    generatingRecoveryCodes: false,
+    generatedRecoveryCodes: false,
+    recoveryCodes: null
+  };
+
+  componentDidMount() {
+    // If we have a TOTP, but no recovery codes for it, immediately go generate some recovery codes.
+    if (this.props.totp && !this.props.totp.recoveryCodes) {
+      this.generateRecoveryCodes();
+    }
+  }
+
+  render() {
+    return (
+      <Dialog isOpen={true} onRequestClose={this.props.onRequestClose} width={540}>
+        <div className="p4">
+          {(this.props.totp.recoveryCodes
+            ? this.renderTotpRecoveryCodes()
+            : this.renderGenerateTotpRecoveryCodes()
+          )}
+        </div>
+      </Dialog>
+    );
+  }
+
+  // TODO: Dry this up...
+  renderGenerateTotpRecoveryCodes() {
```

Comment:
The high level of duplication in `renderGenerateTotpRecoveryCodes` & `renderTotpRecoveryCodes` is not ideal — I think it would be worth maybe refactoring up the dynamic `generateRecoveryCodes` state stuff into a component that lives above this one and then have a simpler UI for actually rendering the dialog content.

Reply:
`renderGenerateTotpRecoveryCodes` no longer needs to exist, as it isn’t possible anymore for a user to have TOTP active but no recovery codes!